README.md
# ngx-gantt

[![npm (scoped)](https://img.shields.io/npm/v/@worktile/gantt?style=flat)](https://www.npmjs.com/package/@ulixe23/gantt)
[![npm](https://img.shields.io/npm/dm/@worktile/gantt)](https://www.npmjs.com/package/@ulixe23/gantt)
![npm bundle size (scoped)](https://img.shields.io/bundlephobia/min/@worktile/gantt)

[coveralls-image]: https://coveralls.io/repos/github/worktile/ngx-gantt/badge.svg?branch=master
[coveralls-url]: https://coveralls.io/github/worktile/ngx-gantt

A modern and powerful gantt component for Angular

## Installation

```bash
$ npm i @ulixe23/gantt html2canvas --save
# or
$ yarn add @ulixe23/gantt html2canvas
```

## Demo

[Try out our live demo](http://gantt.ngnice.com/)

## Usage

### 1. Import the NgxGanttModule to use into your app.module.ts

```ts
import { NgModule } from '@angular/core';
import { NgxGanttModule } from '@ulixe23/gantt';

@NgModule({
  ...
  imports: [ NgxGanttModule, ... ]
  ...
})
export class AppModule {

}
```

### 2. Import style file in angular.json or import style in your style.scss

```json
{
  "styles": ["node_modules/@ulixe23/gantt/styles/index.scss"]
}
```

```scss
@use '@ulixe23/gantt/styles/index.scss';
```

### 3. Using component

component.html

```html
<ngx-gantt #gantt [items]="items">
  <ngx-gantt-table>
    <ngx-gantt-column name="Title" width="300px">
      <ng-template #cell let-item="item"> {{ item.title }} </ng-template>
    </ngx-gantt-column>
  </ngx-gantt-table>
</ngx-gantt>
```

component.ts

```javascript
@Component({
  selector: 'app-gantt-example',
  templateUrl: './gantt.component.html'
})
export class AppGanttExampleComponent {
  items: GanttItem[] = [
    { id: '000000', title: 'Task 0', start: 1627729997, end: 1628421197 },
    { id: '000001', title: 'Task 1', start: 1617361997, end: 1625483597 }
  ];

  constructor() {}
}
```

See [Getting Started](http://gantt.ngnice.com/guides/getting-started) for more details.

## Development

```bash
$ git clone git@github.com:ulixe23/ngx-gantt.git
$ cd ngx-gantt
$ npm ci
$ npm run start
```

## Roadmap

- [ ] virtual scrolling

## LICENSE

[MIT License](https://github.com/worktile/ngx-gantt/blob/master/LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [15.1.4](https://github.com/worktile/ngx-gantt/compare/15.1.3...15.1.4) (2023-08-03)

### Features

- export components ([45c978a](https://github.com/worktile/ngx-gantt/commit/45c978a6fa9a49bbf3fbf144e4b89c8384b903db))
- export components ([5a46f9d](https://github.com/worktile/ngx-gantt/commit/5a46f9d4b3840a1086820e2e315661b992f3af7e))
- remove the redundant white border #INFR-9002 ([a5ecd75](https://github.com/worktile/ngx-gantt/commit/a5ecd75a53501e1147bb784d8e4359ec1edc7cf8)), closes [#INFR-9002](https://github.com/worktile/ngx-gantt/issues/INFR-9002)
- remove the redundant white border at gantt bar #INFR-9002 ([47da959](https://github.com/worktile/ngx-gantt/commit/47da95918919e28e9a108f5b3e94760644fe42ee)), closes [#INFR-9002](https://github.com/worktile/ngx-gantt/issues/INFR-9002)

## [15.1.3](https://github.com/worktile/ngx-gantt/compare/15.1.2...15.1.3) (2023-08-02)

### Bug Fixes

- add bar index when drag #INFR-8617 ([#404](https://github.com/worktile/ngx-gantt/issues/404)) ([a17bf27](https://github.com/worktile/ngx-gantt/commit/a17bf27a6619b75f34920e277263c1d0f7ff6cb9)), closes [#INFR-8617](https://github.com/worktile/ngx-gantt/issues/INFR-8617) [#INFR-8617](https://github.com/worktile/ngx-gantt/issues/INFR-8617)

## [15.1.2](https://github.com/worktile/ngx-gantt/compare/15.1.1...15.1.2) (2023-07-19)

### Bug Fixes

- fix children display error when dropEnterPredicate return false #INFR-8641 ([54c3455](https://github.com/worktile/ngx-gantt/commit/54c3455739e0e6559e18f0e77130c84a50e5e118)), closes [#INFR-8641](https://github.com/worktile/ngx-gantt/issues/INFR-8641)

## [15.1.1](https://github.com/worktile/ngx-gantt/compare/15.1.0...15.1.1) (2023-07-05)

### Bug Fixes

- fix compute scroll bar error ([395d82a](https://github.com/worktile/ngx-gantt/commit/395d82ab1d90c3063367cfb1942d70a844120d42))
- fix misalignment when scrollbar scrolls to the edge ([#389](https://github.com/worktile/ngx-gantt/issues/389)) ([6f5c862](https://github.com/worktile/ngx-gantt/commit/6f5c86281942882e3d115ee55e50212df83d5c24))
- fix two scrollbar error ([b73c3e3](https://github.com/worktile/ngx-gantt/commit/b73c3e30d3eba7dafd567d0ae4b71a12c268ac9d))

# [15.1.0](https://github.com/worktile/ngx-gantt/compare/15.0.0...15.1.0) (2023-05-19)

### chore

- change color [#348](https://github.com/worktile/ngx-gantt/issues/348)fe4 to [#6698](https://github.com/worktile/ngx-gantt/issues/6698)ff #INFR-7105 ([5849ddb](https://github.com/worktile/ngx-gantt/commit/5849ddbe4fc83d941f2dedbeab3fa1505a1b70ad)), closes [#348fe4](https://github.com/worktile/ngx-gantt/issues/348fe4) [#INFR-7105](https://github.com/worktile/ngx-gantt/issues/INFR-7105)

### Features

- auto scroll when dragging #INFR-6189 ([#317](https://github.com/worktile/ngx-gantt/issues/317)) ([d18654d](https://github.com/worktile/ngx-gantt/commit/d18654d3d55b1cd7dd65d38284093ced3eab6cf2)), closes [#INFR-6189](https://github.com/worktile/ngx-gantt/issues/INFR-6189) [#INFR-6189](https://github.com/worktile/ngx-gantt/issues/INFR-6189) [#INFR-6189](https://github.com/worktile/ngx-gantt/issues/INFR-6189)
- **git:** add auto publish ([#364](https://github.com/worktile/ngx-gantt/issues/364)) ([29b2ec6](https://github.com/worktile/ngx-gantt/commit/29b2ec6412f2e7d6d9e30ff70b6ada804a5e4673))
- merge 15.1.0-next ([1aaedf2](https://github.com/worktile/ngx-gantt/commit/1aaedf28b56cf50e1de06ee91121f872f837617d))

### BREAKING CHANGES

- modified some color variables #348fe4 to #6698ff

## [15.1.0-next.12](https://github.com/worktile/ngx-gantt/compare/15.1.0-next.11...15.1.0-next.12) (2023-04-17)

### Bug Fixes

- **gantt:** handle style change when start and end same(#INFR-7407) ([ceca539](https://github.com/worktile/ngx-gantt/commit/ceca539e11a1bdd09f6f089eef1bc7571bce4c09)), closes [#INFR-7407](https://github.com/worktile/ngx-gantt/issues/INFR-7407)

## [15.1.0-next.11](https://github.com/worktile/ngx-gantt/compare/15.1.0-next.10...15.1.0-next.11) (2023-04-13)

### Features

- **gantt:** optimize auto scroll speed: the speed increases when the mouse is closer to the gantt edge #INFR-7196 ([#355](https://github.com/worktile/ngx-gantt/issues/355)) ([da8c377](https://github.com/worktile/ngx-gantt/commit/da8c377e4d48c528dd562bc189ea01590e35e59a)), closes [#INFR-7196](https://github.com/worktile/ngx-gantt/issues/INFR-7196) [#INFR-7196](https://github.com/worktile/ngx-gantt/issues/INFR-7196) [#INFR-7196](https://github.com/worktile/ngx-gantt/issues/INFR-7196) [#INFR-7196](https://github.com/worktile/ngx-gantt/issues/INFR-7196)

## [15.1.0-next.10](https://github.com/worktile/ngx-gantt/compare/15.1.0-next.9...15.1.0-next.10) (2023-04-12)

### Bug Fixes

- **gantt:** auto-scroll should be stopped when start time is greater than end time #INFR-7195 ([#352](https://github.com/worktile/ngx-gantt/issues/352)) ([6f3d29f](https://github.com/worktile/ngx-gantt/commit/6f3d29f8044d27f7ef2534585397609271684647)), closes [#INFR-7195](https://github.com/worktile/ngx-gantt/issues/INFR-7195) [#INFR-7195](https://github.com/worktile/ngx-gantt/issues/INFR-7195) [#INFR-7195](https://github.com/worktile/ngx-gantt/issues/INFR-7195) [#INFR-7195](https://github.com/worktile/ngx-gantt/issues/INFR-7195)
- **gantt:** the fast right (left) dragging start (end) time, the width of the bar is wrong #INFR-7111 ([#351](https://github.com/worktile/ngx-gantt/issues/351)) ([683661c](https://github.com/worktile/ngx-gantt/commit/683661cb3d4b65bee602f17d087871bda9d004b3)), closes [#INFR-7111](https://github.com/worktile/ngx-gantt/issues/INFR-7111) [#INFR-7111](https://github.com/worktile/ngx-gantt/issues/INFR-7111) [#INFR-7111](https://github.com/worktile/ngx-gantt/issues/INFR-7111) [#INFR-7111](https://github.com/worktile/ngx-gantt/issues/INFR-7111)


.docgenirc.js
module.exports = {
    mode: 'full',
    title: 'NgxGantt',
    siteProjectName: 'example',
    outputDir: 'dist/site',
    logoUrl: 'https://cdn.pingcode.com/open-sources/gantt/logo.png',
    repoUrl: 'https://github.com/FabioG95-Ulixe/ngx-gantt',
    defaultLocale: 'it-IT',
    navs: [
        null,
        {
            title: 'Componenti',
            path: 'components',
            locales: {
                'en-us': {
                    title: 'Component'
                }
            }
        },
        {
            title: 'Parametri',
            path: 'configuration',
            lib: 'ngx-gantt',
            locales: {
                'en-us': {
                    title: 'Configuration'
                }
            }
        },
        {
            title: 'GitHub',
            path: 'https://github.com/FabioG95-Ulixe/ngx-gantt',
            isExternal: true
        },
        {
            title: 'Changelog',
            path: '',
            isExternal: true,
            locales: {
                'en-us': {
                    title: 'Changelog'
                }
            }
        }
    ],
    libs: [
        {
            name: 'ngx-gantt',
            rootDir: './example/src/app/configuration',
            exclude: [],
            categories: [
                {
                    id: 'config',
                    title: 'Configurazione',
                    locales: {
                        'en-us': {
                            title: 'Configuration'
                        }
                    }
                }
            ]
        }
    ]
};


example/src/app/components/components.component.ts
import { Component, OnInit } from '@angular/core';
import { DocItem } from '@docgeni/template';

@Component({
    selector: 'app-example-components',
    templateUrl: './components.component.html'
})
export class AppExampleComponentsComponent implements OnInit {
    menus: DocItem[] = [
        {
            id: 'basic',
            title: 'Basic',
            subtitle: '',
            path: 'basic'
        },
        {
            id: 'groups',
            title: 'Groups',
            subtitle: '',
            path: 'groups'
        },
        {
            id: 'virtual-scroll',
            title: 'Virtual Scroll',
            subtitle: '',
            path: 'virtual-scroll'
        },
        {
            id: 'custom-view',
            title: 'Custom View',
            subtitle: '',
            path: 'custom-view'
        },
        {
            id: 'advanced',
            title: 'Advanced',
            subtitle: '',
            path: 'advanced'
        }
    ];

    constructor() {}

    ngOnInit() {}
}


example/src/app/gantt-custom-view/custom-day-view.ts
import {
    GanttView,
    GanttViewOptions,
    primaryDatePointTop,
    secondaryDatePointTop,
    GanttViewDate,
    GanttDate,
    eachDayOfInterval,
    GanttDatePoint,
    GanttViewType
} from 'ngx-gantt';

const viewOptions: GanttViewOptions = {
    cellWidth: 50,
    start: new GanttDate().startOfYear().startOfWeek({ weekStartsOn: 1 }),
    end: new GanttDate().endOfYear().endOfWeek({ weekStartsOn: 1 }),
    addAmount: 1,
    addUnit: 'month',
    fillDays: 1
};

export class GanttViewCustom extends GanttView {
    override showWeekBackdrop = true;

    override showTimeline = true;

    override viewType = GanttViewType.day;

    constructor(start: GanttViewDate, end: GanttViewDate, options?: GanttViewOptions) {
        super(start, end, Object.assign({}, viewOptions, options));
    }

    startOf(date: GanttDate) {
        return date.startOfWeek({ weekStartsOn: 1 });
    }

    endOf(date: GanttDate) {
        return date.endOfWeek({ weekStartsOn: 1 });
    }

    getPrimaryWidth() {
        if (!this.options.showWeekend) {
            return this.getCellWidth() * 5;
        } else {
            return this.getCellWidth() * 7;
        }
    }

    getDayOccupancyWidth(date: GanttDate): number {
        if (!this.options.showWeekend && date.isWeekend()) {
            return 0;
        }
        return this.cellWidth;
    }

    getPrimaryDatePoints(): GanttDatePoint[] {
        const days = eachDayOfInterval({ start: this.start.value, end: this.end.value });
        const points: GanttDatePoint[] = [];
        const dayInWeekMap = {
            '1': 'Lunedì',
            '2': 'Martedì',
            '3': 'Mercoledì',
            '4': 'Giovedì',
            '5': 'Venerdì',
            '6': 'Sabato',
            '0': 'Domenica'
        };
        for (let i = 0; i < days.length; i++) {
            const start = new GanttDate(days[i]);
            const isWeekend = start.isWeekend();
            const point = new GanttDatePoint(
                start,
                `${dayInWeekMap[start.getDay()]}`,
                i * this.getCellWidth() + this.getCellWidth() / 2,
                primaryDatePointTop,
                {
                    isWeekend,
                    isToday: start.isToday()
                }
            );
            if (isWeekend) {
                point.style = { fill: '#ff9f73' };
            }
            if (start.isToday()) {
                point.style = { fill: '#ff9f73' };
            }
            points.push(point);
        }
        if (!this.options.showWeekend) {
            return points
                .filter((point) => !point.additions.isWeekend)
                .map((point, i) => {
                    return { ...point, x: i * this.getCellWidth() + this.getCellWidth() / 2 };
                });
        } else {
            return points;
        }
    }

    getSecondaryDatePoints(): GanttDatePoint[] {
        const days = eachDayOfInterval({ start: this.start.value, end: this.end.value });
        const points: GanttDatePoint[] = [];
        for (let i = 0; i < days.length; i++) {
            const start = new GanttDate(days[i]);
            const isWeekend = start.isWeekend();
            const point = new GanttDatePoint(
                start,
                `${start.format('MM/d')}`,
                i * this.getCellWidth() + this.getCellWidth() / 2,
                secondaryDatePointTop,
                {
                    isWeekend,
                    isToday: start.isToday()
                }
            );
            if (isWeekend) {
                point.style = { fill: '#ff9f73' };
            }
            if (start.isToday()) {
                point.style = { fill: '#ff9f73' };
            }
            points.push(point);
        }

        if (!this.options.showWeekend) {
            return points
                .filter((point) => !point.additions.isWeekend)
                .map((point, i) => {
                    return { ...point, x: i * this.getCellWidth() + this.getCellWidth() / 2 };
                });
        } else {
            return points;
        }
    }
}


example/src/app/gantt-groups/gantt-groups.component.ts
import { Component, OnInit, HostBinding, ViewChild } from '@angular/core';
import { GanttViewType, GanttItem, GanttGroup, NgxGanttComponent } from 'ngx-gantt';
// import { randomGroupsAndItems } from '../helper';
import { of } from 'rxjs';
import { delay } from 'rxjs/operators';
import { random, randomGroupsAndItems, randomItems } from '../helper';

@Component({
    selector: 'app-gantt-groups-example',
    templateUrl: './gantt-groups.component.html'
})
export class AppGanttGroupsExampleComponent implements OnInit {
    views = [
        {
            name: 'Giorno',
            value: GanttViewType.day
        },
        {
            name: 'Settimana',
            value: GanttViewType.week
        },
        {
            name: 'Mese',
            value: GanttViewType.month
        },
        {
            name: 'Trimestre',
            value: GanttViewType.quarter
        },
        {
            name: 'Anno',
            value: GanttViewType.year
        }
    ];

    viewType: GanttViewType = GanttViewType.quarter;

    items: GanttItem[] = [];

    groups: GanttGroup[] = [];

    expanded = true;

    @ViewChild('gantt') ganttComponent: NgxGanttComponent;

    @HostBinding('class.gantt-example-component') class = true;

    constructor() {}

    ngOnInit(): void {
        const { groups, items } = randomGroupsAndItems(10);
        this.groups = groups;
        this.items = items;
    }

    expandAllGroups() {
        if (this.expanded) {
            this.expanded = false;
            this.ganttComponent.collapseAll();
        } else {
            this.expanded = true;
            this.ganttComponent.expandAll();
        }
    }

    childrenResolve = (item: GanttItem) => {
        const children = randomItems(random(1, 5), item);
        return of(children).pipe(delay(1000));
    };
}


example/src/app/gantt/gantt.component.ts
import { Component, OnInit, HostBinding, ViewChild, AfterViewInit } from '@angular/core';
import {
    GanttBarClickEvent,
    GanttViewType,
    GanttDragEvent,
    GanttLineClickEvent,
    GanttLinkDragEvent,
    GanttItem,
    GanttPrintService,
    NgxGanttComponent,
    GanttSelectedEvent,
    GanttBaselineItem,
    GanttView,
    GanttToolbarOptions,
    GanttTableDragEnterPredicateContext,
    GanttTableDragDroppedEvent,
    GanttTableDragStartedEvent,
    GanttTableDragEndedEvent
} from 'ngx-gantt';
import { finalize, of } from 'rxjs';
import { delay } from 'rxjs/operators';
import { ThyNotifyService } from 'ngx-tethys/notify';
import { randomItems, random } from '../helper';

@Component({
    selector: 'app-gantt-example',
    templateUrl: './gantt.component.html',
    providers: [GanttPrintService]
})
export class AppGanttExampleComponent implements OnInit, AfterViewInit {
    views = [
        {
            name: 'Giorno',
            value: GanttViewType.day
        },
        {
            name: 'Settimana',
            value: GanttViewType.week
        },
        {
            name: 'Mese',
            value: GanttViewType.month
        },
        {
            name: 'Quadrimestre',
            value: GanttViewType.quarter
        },
        {
            name: 'Anno',
            value: GanttViewType.year
        }
    ];

    viewType: GanttViewType = GanttViewType.month;

    selectedViewType: GanttViewType = GanttViewType.month;

    isBaselineChecked = false;

    isShowToolbarChecked = true;

    loading = false;

    items: GanttItem[] = [
        { id: '000000', title: 'Task 0', start: 1627729997, end: 1628421197 },
        // { id: '000001', title: 'Task 1', start: 1617361997, end: 1625483597, links: ['000003', '000004', '000000'],  },
        { id: '000001', title: 'Task 1', start: 1617361997, end: 1625483597, links: ['000003', '000004', '0000029'] },
        { id: '000002', title: 'Task 2', start: 1610536397, end: 1610622797, progress: 0.5 },
        { id: '000003', title: 'Task 3 (不可拖动)', start: 1628507597, end: 1633345997, itemDraggable: false },
        { id: '000004', title: 'Task 4', start: 1624705997 },
        { id: '000005', title: 'Task 5', start: 1628075597, end: 1629544397, color: '#709dc1' },
        { id: '000006', title: 'Task 6', start: 1641121997, end: 1645528397 },
        { id: '000007', title: 'Task 7', start: 1639393997, end: 1640862797 },
        { id: '000008', title: 'Task 8', end: 1628783999, color: '#709dc1' },
        { id: '000009', title: 'Task 9', start: 1639307597, end: 1640344397 },
        { id: '0000010', title: 'Task 10', start: 1609067597, end: 1617275597 },
        { id: '0000011', title: 'Task 11', start: 1611918797, end: 1611918797 },
        { id: '0000012', title: 'Task 12', start: 1627816397, end: 1631358797 },
        { id: '0000013', title: 'Task 13', start: 1625051597, end: 1630667597, links: ['0000012'] },
        { id: '0000014', title: 'Task 14', start: 1627920000, end: 1629129599 },
        { id: '0000015', title: 'Task 15', start: 1633259597, end: 1639480397 },
        { id: '0000016', title: 'Task 16', start: 1624965197, end: 1627211597 },
        { id: '0000017', title: 'Task 17', start: 1641035597, end: 1649157197 },
        { id: '0000018', title: 'Task 18', start: 1637061197, end: 1642677197 },
        { id: '0000019', title: 'Task 19', start: 1637925197, end: 1646305997 },
        { id: '0000020', title: 'Task 20', start: 1628334797, end: 1629889997 },
        { id: '0000021', title: 'Task 21', start: 1622891597, end: 1627643597 },
        { id: '0000022', title: 'Task 22', start: 1616238797, end: 1620731597 },
        { id: '0000023', title: 'Task 23', start: 1626693197, end: 1630149197 },
        { id: '0000024', title: 'Task 24', start: 1626174797, end: 1626952397 },
        { id: '0000025', title: 'Task 25', start: 1631013197, end: 1637493197 },
        { id: '0000026', title: 'Task 26', start: 1635937997, end: 1643886797 },
        { id: '0000027', title: 'Task 27', start: 1637665997, end: 1644059597 },
        { id: '0000028', title: 'Task 28', start: 1611400397, end: 1615547597 },
        { id: '0000029', title: 'Task 29', start: 1618053197, end: 1619176397 }
    ];

    toolbarOptions: GanttToolbarOptions = {
        viewTypes: [GanttViewType.day, GanttViewType.month, GanttViewType.year]
    };

    baselineItems: GanttBaselineItem[] = [];

    options = {
        viewType: GanttViewType.day
    };

    viewOptions = {
        dateFormat: {
            month: 'MM'
        }
    };

    @HostBinding('class.gantt-example-component') class = true;

    @ViewChild('gantt') ganttComponent: NgxGanttComponent;

    dropEnterPredicate = (event: GanttTableDragEnterPredicateContext) => {
        return true;
    };

    constructor(private printService: GanttPrintService, private thyNotify: ThyNotifyService) {}

    ngOnInit(): void {
        // init items children
        this.items.forEach((item, index) => {
            if (index % 5 === 0) {
                item.children = randomItems(random(1, 5), item);
            }
        });

        console.log(this.items);
    }

    ngAfterViewInit() {
        setTimeout(() => this.ganttComponent.scrollToDate(1627729997), 200);
    }

    barClick(event: GanttBarClickEvent) {
        this.thyNotify.info('Event: barClick', `Click [${event.item.title}]`);
    }

    lineClick(event: GanttLineClickEvent) {
        this.thyNotify.info('Event: lineClick', `Collegamento da ${event.source.title} a ${event.target.title}`);
    }

    dragMoved(event: GanttDragEvent) {}

    dragEnded(event: GanttDragEvent) {
        this.thyNotify.info('Event: dragEnded', `Drag elemento [${event.item.title}]`);
        this.items = [...this.items];
    }

    selectedChange(event: GanttSelectedEvent) {
        this.thyNotify.info(
            'Event: selectedChange',
            `Selezionati elementi ${(event.selectedValue as GanttItem[]).map((item) => item.id).join('、')}`
        );
    }

    linkDragEnded(event: GanttLinkDragEvent) {
        this.items = [...this.items];
        this.thyNotify.info('Event: linkDragEnded', `Creata associazione`);
    }

    print(name: string) {
        this.printService.print(name);
    }

    scrollToToday() {
        this.ganttComponent.scrollToToday();
    }

    switchChange() {
        if (this.isBaselineChecked) {
            this.baselineItems = [
                { id: '000000', start: 1627728888, end: 1628421197 },
                { id: '000001', start: 1617361997, end: 1625483597 },
                { id: '000002', start: 1610536397, end: 1610622797 },
                { id: '000003', start: 1628507597, end: 1633345997 },
                { id: '000004', start: 1624705997 }
            ];
        } else {
            this.baselineItems = [];
        }
    }

    selectView(type: GanttViewType) {
        this.viewType = type;
        this.selectedViewType = type;
    }

    viewChange(event: GanttView) {
        console.log(event.viewType);
        this.selectedViewType = event.viewType;
    }

    refresh() {
        this.loading = true;
        of(randomItems(30))
            .pipe(
                delay(2000),
                finalize(() => {
                    this.loading = false;
                })
            )
            .subscribe((res) => {
                this.items = res;
            });
    }

    onDragDropped(event: GanttTableDragDroppedEvent) {
        const sourceItems = event.sourceParent?.children || this.items;
        sourceItems.splice(sourceItems.indexOf(event.source), 1);
        if (event.dropPosition === 'inside') {
            event.target.children = [...(event.target.children || []), event.source];
        } else {
            const targetItems = event.targetParent?.children || this.items;
            if (event.dropPosition === 'before') {
                targetItems.splice(targetItems.indexOf(event.target), 0, event.source);
            } else {
                targetItems.splice(targetItems.indexOf(event.target) + 1, 0, event.source);
            }
        }
        this.items = [...this.items];
    }

    onDragStarted(event: GanttTableDragStartedEvent) {
        console.log('拖拽开始了', event);
    }

    onDragEnded(event: GanttTableDragEndedEvent) {
        console.log('拖拽结束了', event);
    }
}
